Add option for flexible space on material SearchDelegate

HenriqueNas opened the pull request:
This pull request introduces the `buildFlexibleSpace` method to the `SearchDelegate` class in the material library. 
It allows users to add a flexible space widget to the `AppBar` in a `_SearchPage`, providing more customization options.

This PR does not fix any specific issue as there are no open issues related to this feature.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

HenriqueNas committed:
feat(material): :sparkles: Add buildFlexibleSpace method to material SearchDelegate class

introduce the option for add a flexibleSpace Widget to AppBar in a _SearchPage

HenriqueNas committed:
test(material tests): :white_check_mark: update tests for search widget

HenriqueNas committed:
Merge branch 'flutter:master' into master

HenriqueNas committed:
test(material test): :white_check_mark: add test for buildFlexibleSpace usage on SearchDelegate and SearchPage classes

HenriqueNas committed:
Merge branch 'master' into master

HenriqueNas committed:
Merge branch 'master' into master

HenriqueNas committed:
Merge branch 'master' into master

goderbauer commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -211,6 +211,16 @@ abstract class SearchDelegate<T> {
+  /// See also:
+  ///
+  ///  * [AppBar.flexibleSpace], the intended use for the return value of this method.
+  ///
```
nit: remove blank line.

HenriqueNas replied:
done ✅

goderbauer commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -211,6 +211,16 @@ abstract class SearchDelegate<T> {
   ///
   PreferredSizeWidget? buildBottom(BuildContext context) => null;
 
+  /// Widget to display in the [AppBar] flexible space.
```
Maybe just:

```suggestion
  /// Widget to display in the [AppBar.flexibleSpace].
```

and then remove the "see also" section?

HenriqueNas replied:
I did this way to keep the pattern of others documentation for build methods in the class.
You can see it in `PreferredSizeWidget? buildBottom(BuildContext)` and `List<Widget>? buildActions(BuildContext)`.

changes:
```dart
  /// Widget to display a flexible space in the [AppBar].
  ///
  /// Returns null by default, i.e. a flexible space widget is not included.
  ///
  /// See also:
  ///
  ///  * [AppBar.flexibleSpace], the intended use for the return value of this method.
  Widget? buildFlexibleSpace(BuildContext context) => null;
```

goderbauer commented on packages/flutter/test/material/search_test.dart:
```diff
@@ -1079,6 +1112,9 @@ class _TestEmptySearchDelegate extends SearchDelegate<String> {
     return const Text('Results');
   }
 
+  // @override
+  // Widget? buildFlexibleSpace(BuildContext context) => null;
```
remove commented out code?

HenriqueNas replied:
done ✅

goderbauer commented on packages/flutter/test/material/search_test.dart:
```diff
@@ -980,6 +1006,7 @@ class _TestSearchDelegate extends SearchDelegate<String> {
     this.suggestions = 'Suggestions',
     this.result = 'Result',
     this.actions = const <Widget>[],
+    this.flexibleSpace = const Text('FlexibleSpace'),
```
make this null by default so this change doesn't invalidate all the other tests and make sure everything still works without this set? Then, add specific tests that set this to non-null and check that everything works.

HenriqueNas replied:
done ✅

HenriqueNas committed:
refactor: :white_check_mark: update test following code review suggestions

set flexibleSpace parameter default value as null (_TestSearchDelegate material search test)

HenriqueNas committed:
Merge branch 'master' into master

HenriqueNas committed:
Merge branch 'master' into master

HenriqueNas committed:
Merge branch 'master' into master

HenriqueNas committed:
Merge branch 'master' into master

HenriqueNas committed:
Merge branch 'master' into master

HenriqueNas committed:
Merge branch 'master' into master

HenriqueNas committed:
Merge branch 'master' into master

HenriqueNas commented:
Hey @goderbauer, I think I did everything that was needed here.
Could you rereview this PR, and let me know if have something more I can do ?

HenriqueNas committed:
Merge branch 'master' into master

goderbauer reviewed (approved):
LGTM

chunhtai reviewed (approved):
LGTM

HenriqueNas committed:
Merge branch 'master' into master
